### ISSUE TITLE
amd64 implementations clobber the frame pointer register

### ISSUE THREAD

**Issue by nsrip-dd** [dgryski/go-metro#9]
The amd64 implementations of Hash64 and Hash64Str clobber the frame pointer register (BP). They return to their caller leaving an invalid frame pointer value in the BP register ([this one specifically](https://github.com/dgryski/go-metro/blob/adc40b04c140ccfcbb2d243f40ca26a196eccd0d/metro_amd64.s#L117)). This breaks frame pointer unwinding, which the Go runtime uses to collect call stacks for the execution tracer and some of the profilers.

If BP is going to be used as a scratch register, it should at least be saved before use and restored on exit. e.g. PUSHQ BP before the first use and POPQ BP before returning.

**Comment by lizthegrey** [comment]
This happens because in peachpy rbp and rsp are included in [GeneralPurposeRegister64](https://github.com/Maratyszcza/PeachPy/blob/349e8f836142b2ed0efeb6bb99b1b715d87202e9/peachpy/x86_64/registers.py#L242).

```
rsp = GeneralPurposeRegister64(4)
rbp = GeneralPurposeRegister64(5)
```

the asm generation just trusts that any "general purpose" register it's handed is safe to clobber.

**Comment by nsrip-dd** [comment]
Indeed. This is a point in favor of moving from peachpy to Avo (#7) since it knows to save the frame pointer register: https://github.com/mmcloughlin/avo/blob/e2b6ba5dd95e8cf47eeafb3b56794952983cae6d/pass/reg.go#L171 :)

**Comment by nsrip-dd** [comment]
@lizthegrey I sent https://github.com/dgryski/go-metro/pull/10 to port the assembly generation to Avo, which fixes this issue
